packages/valory/skills/twitter_scoring_abci/behaviours.py

```diff
@@ -130,6 +137,103 @@ def _check_daily_limit(self) -> Tuple:
         return False, number_of_tweets_pulled_today, last_tweet_pull_window_reset
 
 
+class TwitterRandomnessBehaviour(RandomnessBehaviour):
+    """Retrieve randomness."""
+
+    matching_round = TwitterRandomnessRound
+    payload_class = TwitterRandomnessPayload
+
+
+class TwitterSelectKeepersBehaviour(TwitterScoringBaseBehaviour):
+    """Select the keeper agent."""
+
+    matching_round = TwitterSelectKeepersRound
+    payload_class = TwitterSelectKeepersPayload
+
+    def _select_keepers(self) -> List[str]:
+        """
+        Select new keepers randomly.
+
+        1. Sort the list of participants who are not blacklisted as keepers.
+        2. Randomly shuffle it.
+        3. Pick the first keepers in order.
+        4. If they have already been selected, pick the next ones.
+
+        :return: the selected keepers' addresses.
+        """
+        # Get all the participants who have not been blacklisted as keepers
+        non_blacklisted = (
+            self.synchronized_data.participants
+            - self.synchronized_data.blacklisted_keepers
+        )
+        if not non_blacklisted:
+            raise RuntimeError(
+                "Cannot continue if all the keepers have been blacklisted!"
+            )
+
+        # Sorted list of participants who are not blacklisted as keepers
+        relevant_set = sorted(list(non_blacklisted))
+
+        needed_keepers = math.ceil(
+            self.synchronized_data.nb_participants / 2
+        )  # half or 1
+
+        # Check if we need random selection
+        if len(relevant_set) <= needed_keepers:
+            keeper_addresses = list(relevant_set)
+            self.context.logger.info(f"Selected new keepers: {keeper_addresses}.")
+            return keeper_addresses
+
+        # Random seeding and shuffling of the set
+        random.seed(self.synchronized_data.keeper_randomness)
+        random.shuffle(relevant_set)
+
+        # If the keeper is not set yet, pick the first address
+        keeper_addresses = relevant_set[0:2]
+
+        # If the keepers have been already set, select the next ones.
+        if (
+            self.synchronized_data.are_keepers_set
+            and len(self.synchronized_data.participants) > 2
+        ):
+            old_keeper_index = relevant_set.index(
+                self.synchronized_data.most_voted_keeper_addresses[0]
+            )
+            keeper_addresses = [
+                relevant_set[
+                    (old_keeper_index + 2) % len(relevant_set)
+                ],  # skip the previous 2
+                relevant_set[(old_keeper_index + 3) % len(relevant_set)],
+            ]
```

Comment:
This looks a bit odd to me, we're slicing the keepers from the set of relevant keepers and after that we are checking if the keepers are set and using the index of the existing keepers to select the next ones. The problem here is that we're shuffling the keepers which means the relevant_set is none deterministic and we cannot use indexes to select different keepers. I would suggest we loop through the relevant set and select keepers which are not in `self.synchronized_data.most_voted_keeper_addresses`

Reply:
This follows the current implementation for single keeper selection, which is tested code, but I'd like to have @Adamantios opinion.

Reply:
This is deterministic across the agents because they are using the same seed when shuffling. 
This was implemented in this manner to ensure that we consistently choose all the agents as keepers.